### PR TITLE
refactor: :label: stronger types for `StateValue` when using Typegen

### PR DESCRIPTION
**Pull request by brunocrosier**
Demo of `state.value` having loose types while using Typegen: https://codesandbox.io/s/github/brunocrosier/test-xstate-typegen

image

Borrowing the type from: https://github.com/statelyai/xstate/blob/main/packages/core/src/State.ts#L321

### CODE REVIEW

**Comment by brunocrosier** [comment]
> lol @erikras just literally told me he wanted this when we went out for sushi

jealous! 🍣 

I'm hoping this will work..!

**Comment by Andarist** [comment]
I **think**, but that would have to be verified, that reusing `matchesStates` is not good enough here because it accepts some object~ flavors that are not correct for the `state.value`.

q: why do you want to use `state.value` in the first place? I'm not saying that it's not OK - I just want to understand the context behind the request etc. I think that `state.value` & `state.matches` are so close to each other that I wonder if there is a value in having both of them in the API.

File: packages/core/src/StateNode.ts [hunk 1]
```diff
@@ -780,8 +782,8 @@ class StateNode<
 
     return next;
   }
-  private transitionCompoundNode(
-    stateValue: StateValueMap,
+  private transitionCompoundNode<TResolvedTypesMeta>(
```
[the remark this answers is not in the document]
Comment (brunocrosier) [reply packages/core/src/StateNode.ts]: you're 100% correct - great catch! should be fixed here: https://github.com/statelyai/xstate/pull/3342/commits/b2a466f2ba19eef4a0a26c31a3897423fd23a4e2

**Comment by brunocrosier** [comment]
> I **think**, but that would have to be verified, that reusing `matchesStates` is not good enough here because it accepts some object~ flavors that are not correct for the `state.value`.

Sure thing! I could create a PR to add something like `stateValues` here: https://github.com/statelyai/xstate-tools/blob/e2d7a2b498fb4d3a666c434748d678ce6ca5f8e7/packages/shared/src/getTypegenOutput.ts#L164 ?

What would that look like though? How would it be different to `matchesStates` ?

> q: why do you want to use `state.value` in the first place? I'm not saying that it's not OK - I just want to understand the context behind the request etc. I think that `state.value` & `state.matches` are so close to each other that I wonder if there is a value in having both of them in the API.

Basically, just for better type safety. I created a small repro here to show an example of when we get a type error when we should not: https://codesandbox.io/s/github/brunocrosier/test-xstate-typegen

Happy to explain in more detail if that's unclear!

**Comment by Andarist** [comment]
> What would that look like though? How would it be different to matchesStates ?

One important difference is that only the "object syntax" is valid for `state.value` (+ string values for top-level atomic state) but a "dotted" notation is not valid:
https://github.com/statelyai/xstate-tools/blob/041846a873d58247fe17206f1e6149916203b449/packages/shared/src/getTypegenOutput.ts#L93-L97

Another one is that `.matches` always mark properties as optional:
https://github.com/statelyai/xstate-tools/blob/041846a873d58247fe17206f1e6149916203b449/packages/shared/src/getStateMatchesObjectSyntax.ts#L24
but for parallel regions in the `.value` this probably wouldn't be true, those should actually always be required.

**Comment by brunocrosier** [comment]
OK, I will see if I can create a PR in `xstate-tools` to add  `stateValues` with the requirements that you mentioned 🙂

**Comment by Andarist** [comment]
It would still be pretty interesting to see some real-life examples of how you are using `state.value` in your code.

**Comment by brunocrosier** [comment]
> It would still be pretty interesting to see some real-life examples of how you are using `state.value` in your code.

Unfortunately I don't have any examples apart from that minimal CodeSandbox, sorry! So far I've only played around with XState in personal projects - but I guess that stronger types are always better, right? 

Maybe @erikras has some examples of potential use-cases?

Added PR here: https://github.com/statelyai/xstate-tools/pull/160

**Comment by davidkpiano** [comment]
Superseded by #4498

**Pull request closed without merging**